Documentation/design-docs/standalone-gc-eventing.md

```diff
@@ -0,0 +1,212 @@
+`System.Diagnostics.Tracing.EventSource`. Native code (i.e. the runtime) issues events by calling macros
+that delegate the issuing of events to autogenerated code that is generated to interface with the
+underlying event implementation. Events are only issued if they are turned on; the mechanism by which
+events are turned on is not in the scope of this document
+* The **payload** of an event is some amount of data that is delivered with the event itself. Its size
+may be variable. Most events that are fired by the runtime have a schema (predefined layout), but it
+is not a requirement.
+
+## Goals
+
+The goal of this document is to describe a system that allows for the efficient firing of
+performance events by a standalone GC. This system must have two properties in order to be
+acceptable:
+
+1. It must be *efficient* to query whether or not a particular event is turned on. It is not acceptable
+to perform an indirection (i.e. cross the GC/EE interface boundary) in order to get this information.
+2. The cost of firing an event by a standalone GC should be comparable to the cost of firing an event
+without using a standalone GC.
+
+## Querying Whether Events Are Enabled
+
+Tt is not acceptable to perform an indirection when querying whether or not a requested event
+is enabled. Therefore, it follows that the GC must maintain some state about what events are currently 
+enabled. Events are enabled through *keywords* and *levels*; a particular event is enabled if both the 
+event's keyword and the event's level have been enabled on the provider to which the event belongs.
+
+To accomplish this, the GC will contain a class with this signature:
+
+```c++
+// Constructed on GC startup for every event that the GC can fire.
+struct GCEvent
+{
+    /* opaque, but will contain information about what event this is,
+       its keyword, its level, and a handle to the EE's representation of this event */
+    /* this structure is potentially auto-generated from an event manifest */
+};
+
+class GCEventStatus
+{
+public:
+    // Returns true if this event is currently enabled, false otherwise.
+    static bool IsEventEnabled(const GCEvent& event);
+
+    // Enables events with the given keyword and level.
+    static void Enable(int keyword, int level);
+
+    // Disables events the given keyword and level.
+    static void Disable(int keyword, int level);
+};
+
+```
+
+The GC will use `GCEventStatus::IsEventEnabled` to query whether or not a particular event is enabled.
+Whenever the EE observes a change in what keywords or levels are enabled, it must inform the GC of the
+change so that it can update `GCEventStatus` using `Enable` and `Disable`. The exact mechanism by which
+the EE observes a change in the event state is described further below. ("Getting Informed of Changes to 
+Event State")
+
+When the EE *does* observe a change, it informs the GC using these two new methods on `IGCHeap`:
+
+```c++
+void IGCHeap::EnableEvents(int keyword, int level)
```

Comment:
You are probably going to need to know the provider as well since there are multiple providers that the GC emits events to.

Reply:
good point!